File: csharp-playwright-framework/PlaywrightFramework/Pages/ProductPage.cs
using Microsoft.Playwright;

namespace PlaywrightFramework.Pages;

/// <summary>
/// Page Object for Product/Shopping page
/// </summary>
public class ProductPage : BasePage
{
    // Locators
    private readonly ILocator _productCards;
    private readonly ILocator _addToCartButtons;
    private readonly ILocator _cartIcon;
    private readonly ILocator _cartCount;
    private readonly ILocator _filterDropdown;
    private readonly ILocator _sortDropdown;
    private readonly ILocator _searchBox;
    private readonly ILocator _priceLabels;

    public ProductPage(IPage page) : base(page)
    {
        _productCards = page.Locator(".product-card, [data-testid='product']");
        _addToCartButtons = page.Locator("button:has-text('Add to Cart'), [data-testid='add-to-cart']");
        _cartIcon = page.Locator(".cart-icon, [data-testid='cart-icon']");
        _cartCount = page.Locator(".cart-count, [data-testid='cart-count']");
        _filterDropdown = page.Locator("#filter, [data-testid='filter-dropdown']");
        _sortDropdown = page.Locator("#sort, [data-testid='sort-dropdown']");
        _searchBox = page.Locator("input[placeholder*='Search'], [data-testid='product-search']");
        _priceLabels = page.Locator(".product-price, [data-testid='product-price']");
    }

    /// <summary>
    /// Navigate to products page
    /// </summary>
    public async Task GotoAsync(string url)
    {
        await NavigateAsync(url);
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Check if products are displayed
    /// </summary>
    public async Task<bool> AreProductsDisplayedAsync()
    {
        return await _productCards.First.IsVisibleAsync();
    }

    /// <summary>
    /// Get count of displayed products
    /// </summary>
    public async Task<int> GetProductCountAsync()
    {
        return await _productCards.CountAsync();
    }

    /// <summary>
    /// Add first product to cart
    /// </summary>
    public async Task AddFirstProductToCartAsync()
    {
        await _addToCartButtons.First.ClickAsync();
        await WaitAsync(500); // Wait for cart update animation
    }

    /// <summary>
    /// Add product to cart by index (0-based)
    /// </summary>
    public async Task AddProductToCartByIndexAsync(int index)
    {
        await _addToCartButtons.Nth(index).ClickAsync();
        await WaitAsync(500);
    }

    /// <summary>
    /// Add product to cart by name
    /// </summary>
    public async Task AddProductToCartByNameAsync(string productName)
    {
        var product = Page.Locator($"text={productName}").First;
        var addButton = product.Locator("..").Locator("button:has-text('Add to Cart')").First;
        await addButton.ClickAsync();
        await WaitAsync(500);
    }

    /// <summary>
    /// Get cart item count
    /// </summary>
    public async Task<int> GetCartCountAsync()
    {
        try
        {
            var countText = await _cartCount.TextContentAsync();
            return int.Parse(countText ?? "0");
        }
        catch
        {
            return 0;
        }
    }

    /// <summary>
    /// Click cart icon to view cart
    /// </summary>
    public async Task ClickCartAsync()
    {
        await _cartIcon.ClickAsync();
    }

    /// <summary>
    /// Filter products by category
    /// </summary>
    public async Task FilterByCategoryAsync(string category)
    {
        await _filterDropdown.SelectOptionAsync(category);
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Sort products
    /// </summary>
    public async Task SortProductsAsync(string sortOption)
    {
        await _sortDropdown.SelectOptionAsync(sortOption);
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Search for products
    /// </summary>
    public async Task SearchProductsAsync(string searchTerm)
    {
        await _searchBox.FillAsync(searchTerm);
        await _searchBox.PressAsync("Enter");
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    /// <summary>
    /// Get list of product prices
    /// </summary>
    public async Task<List<decimal>> GetProductPricesAsync()
    {
        var prices = new List<decimal>();
        var count = await _priceLabels.CountAsync();

        for (int i = 0; i < count; i++)
        {
            var priceText = await _priceLabels.Nth(i).TextContentAsync();
            if (priceText != null)
            {
                // Remove currency symbols and parse
                var cleanPrice = priceText.Replace("$", "").Replace(",", "").Trim();
                if (decimal.TryParse(cleanPrice, out var price))
                {
                    prices.Add(price);
                }
            }
        }

        return prices;
    }

    /// <summary>
    /// Get product name by index
    /// </summary>
    public async Task<string> GetProductNameAsync(int index)
    {
        var productCard = _productCards.Nth(index);
        var nameElement = productCard.Locator(".product-name, h3, [data-testid='product-name']").First;
        return await nameElement.TextContentAsync() ?? "";
    }

    /// <summary>
    /// Verify product is in stock
    /// </summary>
    public async Task<bool> IsProductInStockAsync(int index)
    {
        var productCard = _productCards.Nth(index);
        var outOfStockLabel = productCard.Locator("text=/out of stock/i");
        return !await outOfStockLabel.IsVisibleAsync();
    }

    /// <summary>
    /// Add multiple products to cart
    /// </summary>
    public async Task AddMultipleProductsToCartAsync(int count)
    {
        var availableProducts = Math.Min(count, await _addToCartButtons.CountAsync());

        for (int i = 0; i < availableProducts; i++)
        {
            await AddProductToCartByIndexAsync(i);
            await WaitAsync(300); // Small delay between additions
        }
    }
}

File: python-playwright-framework/pages/product_page.py
"""
Product Page Object Model
Contains all elements and actions for product/catalog pages
"""

from playwright.sync_api import Page, Locator
from pages.base_page import BasePage


class ProductPage(BasePage):
    """Product page object model"""
    
    # Page URL
    PRODUCTS_URL = "/products"
    
    def __init__(self, page: Page):
        super().__init__(page)
        
        # Initialize locators
        self._product_list = page.locator('.products-list, .product-grid, [data-test="products"]')
        self._product_cards = page.locator('.product-card, .product-item, [data-test="product"]')
        self._add_to_cart_button = page.locator('button:has-text("Add to Cart"), [data-test="add-to-cart"]')
        self._cart_icon = page.locator('.cart-icon, [data-test="shopping-cart"], a:has-text("Cart")')
        self._cart_count = page.locator('.cart-count, .cart-badge, [data-test="cart-count"]')
        self._product_title = page.locator('.product-title, .product-name, h2')
        self._product_price = page.locator('.product-price, .price, [data-test="price"]')
        self._product_image = page.locator('.product-image, img[alt*="product"]')
        self._filter_dropdown = page.locator('select[name="filter"], [data-test="filter"]')
        self._sort_dropdown = page.locator('select[name="sort"], [data-test="sort"]')
        self._search_input = page.locator('input[placeholder*="Search products"]')
    
    def goto(self) -> None:
        """Navigate to products page"""
        self.navigate(self.PRODUCTS_URL)
        self.wait_for_page_load()
    
    def get_product_count(self) -> int:
        """Get total number of products"""
        return self.get_count(self._product_cards)
    
    def click_product(self, product_name: str) -> None:
        """Click on a specific product by name"""
        product = self.page.locator(f'.product-card:has-text("{product_name}")')
        self.click(product)
    
    def add_product_to_cart(self, product_name: str) -> None:
        """Add product to cart by name"""
        add_to_cart_btn = self.page.locator(
            f'.product-card:has-text("{product_name}") button:has-text("Add to Cart")'
        )
        self.click(add_to_cart_btn)
    
    def get_cart_count(self) -> int:
        """Get cart item count"""
        try:
            count_text = self.get_text(self._cart_count)
            return int(count_text) if count_text else 0
        except:
            return 0
    
    def open_cart(self) -> None:
        """Open shopping cart"""
        self.click(self._cart_icon)
    
    def get_all_product_titles(self) -> list[str]:
        """Get all product titles"""
        return self.get_all_text(self._product_title)
    
    def get_all_product_prices(self) -> list[str]:
        """Get all product prices"""
        return self.get_all_text(self._product_price)
    
    def filter_by_category(self, category: str) -> None:
        """Filter products by category"""
        self.select_by_label(self._filter_dropdown, category)
        self.wait_for_page_load()
    
    def sort_products(self, sort_option: str) -> None:
        """Sort products"""
        self.select_by_label(self._sort_dropdown, sort_option)
        self.wait_for_page_load()
    
    def search_products(self, query: str) -> None:
        """Search for products"""
        self.fill(self._search_input, query)
        self.press_key('Enter')
        self.wait_for_page_load()
    
    def are_products_displayed(self) -> bool:
        """Check if products are displayed"""
        return self.is_visible(self._product_list)
